testkit/src/mill/testkit/BspServerUtil.scala

```diff
@@ -0,0 +1,184 @@
+    def onBuildTaskFinish(params: b.TaskFinishParams): Unit = ()
+    def onBuildTaskProgress(params: b.TaskProgressParams): Unit = ()
+    def onBuildTaskStart(params: b.TaskStartParams): Unit = ()
+    def onRunPrintStderr(params: b.PrintParams): Unit = ()
+    def onRunPrintStdout(params: b.PrintParams): Unit = ()
+  }
+
+  lazy val dummyBuildClient: b.BuildClient = new DummyBuildClient {}
+
+  val gson: Gson = new GsonBuilder().setPrettyPrinting().create()
+  def compareWithGsonFixture[T: ClassTag](
+      value: T,
+      fixturePath: os.Path,
+      replaceAll: Seq[(String, String)] = Nil
+  ): Unit = {
+
+    def doReplaceAll(input: String, inverse: Boolean = false): String =
+      replaceAll.foldLeft(input) {
+        case (input0, (from0, to0)) =>
+          val (from, to) = if (inverse) (to0, from0) else (from0, to0)
+          input0.replace(from, to)
+      }
+
+    val exists = os.exists(fixturePath)
+    val readOpt = Option.when(exists) {
+      val bytes = os.read.bytes(fixturePath)
+      gson.fromJson(
+        doReplaceAll(new String(bytes), inverse = true),
+        implicitly[ClassTag[T]].runtimeClass
+      )
+    }
+
+    if (!readOpt.contains(value))
+      if (updateFixtures) {
+        System.err.println(if (exists) s"Updating $fixturePath" else s"Writing $fixturePath")
+        val jsonStr = gson.toJson(
+          value,
+          implicitly[ClassTag[T]].runtimeClass
+        )
+        os.write.over(fixturePath, doReplaceAll(jsonStr), createFolders = true)
+      } else
+        Predef.assert(
+          false,
+          if (exists) s"Error: value differs from fixture at $fixturePath"
+          else s"Error: no fixture found at $fixturePath"
+        )
+    else if (updateFixtures) {
+      // Fixture on disk might need to be updated anyway, if replaceAll changed
+      // and new strings should be replaced
+      val jsonStr = doReplaceAll(
+        gson.toJson(
+          value,
+          implicitly[ClassTag[T]].runtimeClass
+        )
+      )
+      val onDiskJsonStr = os.read(fixturePath)
+      if (jsonStr != onDiskJsonStr) {
+        System.err.println(s"Updating $fixturePath")
+        os.write.over(fixturePath, jsonStr)
+      }
+    }
+  }
+
+  val bspJsonrpcPool: ExecutorService = Executors.newCachedThreadPool(
+    new ThreadFactory {
+      val counter = new AtomicInteger
+      def newThread(runnable: Runnable): Thread = {
+        val t = new Thread(runnable, s"mill-bsp-integration-${counter.incrementAndGet()}")
+        t.setDaemon(true)
+        t
+      }
+    }
+  )
+
+  trait MillBuildServer extends b.BuildServer with b.JvmBuildServer
+      with b.JavaBuildServer with b.ScalaBuildServer
+
+  def withBspServer[T](
+      workspacePath: os.Path,
+      millTestSuiteEnv: Map[String, String],
+      client: b.BuildClient = dummyBuildClient
+  )(f: (MillBuildServer, b.InitializeBuildResult) => T): T = {
+
+    val jsonFile = workspacePath / Constants.bspDir / s"${Constants.serverName}.json"
```

Comment:
`jsonFile` needs a better name. `serverMetadataFile`?

Reply:
Renamed it to `bspMetadataFile` in the latest push